File: original artifact/Ian_Foss_cs-360_project_three_inventory_app/cs360_project_three/app/src/main/java/com/example/cs_360_project_three/UserRepository.java
package com.example.cs_360_project_three;

import android.content.Context;

import androidx.lifecycle.LiveData;

import java.util.concurrent.Executor;
import java.util.concurrent.Executors;
import java.util.function.Consumer;

public class UserRepository {

    private final UserDao userDao;
    private final Executor ioExecutor = Executors.newSingleThreadExecutor();


    public UserRepository(Context context) {
        InventoryDatabase db = InventoryDatabase.getInstance(context);
        userDao = db.userDao();
    }


    public void insert(UserEntity user) {
        ioExecutor.execute(() -> userDao.insertUser(user));
    }

    public void update(UserEntity user) {
        ioExecutor.execute(() -> userDao.updateUser(user));
    }


    // Gets user on background thread
    public void getUser(String username, Consumer<UserEntity> callback) {
        ioExecutor.execute(() -> {
            UserEntity u = userDao.getUserByUsername(username);
            callback.accept(u);
        });
    }

    // Gets user by username and password to validate if account exists
    public void validateUser(String username, String password, Consumer<UserEntity> callback) {
        ioExecutor.execute(() -> {
            UserEntity user = userDao.getUserByCredentials(username, password);
            callback.accept(user);  // null if not found
        });
    }

}

File: enhancements/enhancement-three/Enhancement Three/app/src/main/java/com/example/CS499_Capstone/UserRepository.kt
package com.example.CS499_Capstone

import android.content.Context
import java.util.concurrent.Executor
import java.util.concurrent.Executors
import java.util.function.Consumer

class UserRepository(context: Context) {
    private val userDao: UserDao
    private val ioExecutor: Executor = Executors.newSingleThreadExecutor()


    init {
        val db = InventoryDatabase.getInstance(context)
        userDao = db.userDao()
    }


    fun insert(user: UserEntity) {
        ioExecutor.execute(Runnable { userDao.insertUser(user) })
    }

    fun update(user: UserEntity) {
        ioExecutor.execute(Runnable { userDao.updateUser(user) })
    }

    fun usernameExists(username: String, callback: (Boolean) -> Unit) {

        ioExecutor.execute {
            val exists = userDao.usernameExists(username) > 0
            callback(exists)
        }
    }


    // Gets user on background thread
    fun getUser(username: String, callback: (UserEntity?) -> Unit) {
        ioExecutor.execute(Runnable {
            val u = userDao.getUserByUsername(username)
            callback(u)
        })
    }

    // Gets user by username and password to validate if account exists
    fun validateUser(username: String, password: String, callback: (UserEntity?) -> Unit) {
        ioExecutor.execute(Runnable {
            val user = userDao.getUserByUsername(username)

            if (user == null) {
                callback(user) // null if not found
                return@Runnable
            }
            val isValidPassword = PasswordUtils.verifyPassword(
                password,
                user.salt,
                user.password
            )
            if (isValidPassword) {
                callback(user)
            }
            else {
                callback(null)
            }
        })
    }
}
